File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    void SumUntial(vector<int>& candidates,vector<vector<int>>& ans,vector<int>& combination,int target,int left)
    {
        if(target==0)
        {
            ans.push_back(combination);
            return;
        }
        for(int i = left; i < candidates.size(); i++)
        {
            if(target >= candidates[i])
            {
                combination.push_back(candidates[i]);
                SumUntial(candidates,ans,combination,target-candidates[i],i);
                combination.pop_back();
            }
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        
        vector<vector<int>> ans;
        vector<int> combination;
        
        SumUntial(candidates,ans,combination,target,0);
        return ans;        
    }
};

File: 39-combination-sum/39-combination-sum.java
class Solution {
    public void dfs(int[] arr, List<List<Integer>> ans, List<Integer> comb, int target,int idx)
    {
        if(target == 0)
        {
            ans.add(new ArrayList<>(comb));
            return;
        }
        
        for(int i = idx; i < arr.length; i++)
        {
            if(target >= arr[i])
            {
                comb.add(arr[i]);
                dfs(arr, ans, comb ,target - arr[i], i);
                comb.remove(comb.size()-1);
            }
        }
    }
    public List<List<Integer>> combinationSum(int[] arr, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        dfs(arr, ans,new ArrayList<>() ,target,0);
        return ans;
        
    }
}